modules/javafx.base/src/main/java/javafx/beans/value/ObservableValue.java

```diff
@@ -251,4 +252,63 @@ default ObservableValue<T> orElse(T constant) {
     default <U> ObservableValue<U> flatMap(Function<? super T, ? extends ObservableValue<? extends U>> mapper) {
         return new FlatMappedBinding<>(this, mapper);
     }
+
+    /**
+     * Returns an {@code ObservableValue} that holds this value whenever the given
+     * condition evaluates to {@code true}, otherwise holds the last value when
+     * {@code condition} became {@code false}. The value is updated whenever this
+     * {@code ObservableValue} changes, unless the condition currently evaluates
+     * to {@code false}.
+     * <p>
+     * The returned {@code ObservableValue} only observes this value when the given
+     * {@code condition} evaluates to {@code true}. This allows this {@code ObservableValue}
+     * and the conditional {@code ObservableValue} to be garbage collected if neither is
+     * otherwise strongly referenced when {@code condition} becomes {@code false}.
```

Comment:
What happens if they are GC'd and the conditional becomes `true` later?

Reply:
If I have a (dumb) method

```java
void someMethod(ObservableValue<String> longLivedProperty) {
    ObservableValue<Boolean> condition = new SimpleBooleanProperty(true);
    ObservableValue<String> whenProperty = longLivedProperty.when(condition)
}
```

then shouldn't `condition` and `whenProperty` be eligible for GC even when `condition` holds `true`? If not, do I not get a memory leak because I can't change `condition` to `false` to allow garbage collection?

Reply:
Well, this specific example, yes, they'll be both eligible for GC, but that's because you're not observing `whenProperty` anywhere. When not observed, it won't create a listener on `longLivedProperty`.

Assuming that `whenProperty` is observed and with `condition` being always `true` it is basically the same as:

     ObservableValue<String> whenProperty = longLivedProperty.map(x -> x);

As long as `whenProperty` is observed, `longLivedProperty` has a direct reference to it (via listener).  With `when` however, the `condition` can be used to (temporarily or permanently) break this reference. You could set it to `false` in `dispose` for example, triggering the unregistration of the listener on `longLivedProperty` (for multiple properties at once if desired):

      ObservableValue<Boolean> active = new SimpleBooleanProperty(true);

      MySkin() {
           getSkinnable().textProperty().when(active).addListener(this::updateSkinStuffs);
           getSkinnable().fontProperty().when(active).addListener(this::updateSkinStuffs);
           getSkinnable().wrapTextProperty().when(active).addListener(this::updateSkinStuffs);
      }

      void dispose() {
           active.set(false);  // kills all listeners/bindings/etc in a single line
      }

Or the other case I really like:

     label.textProperty().bind(longLivedProperty.when(label::isShownProperty));

As long as `label` is showing, it takes updates from `longLivedProperty`; if the screen is closed, it detaches, and (assuming all such bindings are nicely detached) the label (and the rest of the UI it may be part of) can be GC'd without problem.

If you kept a reference to the UI (containing that label), then you also kept a reference to the label, to its text property, and to that binding with then `when` in it -- making that UI visible again restores the listener on `longLivedProperty` and the label immediately receives whatever the latest value is in that property if it changed while the UI was invisible.

Reply:
@nlisker is this adequate, or should the docs reflect this better?